src/OptimalTransport.jl

```diff
@@ -785,4 +786,68 @@ function quadreg(mu, nu, C, ϵ; θ=0.1, tol=1e-5, maxiter=50, κ=0.5, δ=1e-5)
     return sparse(γ')
 end
 
+"""
+    sinkhorn_divergence(c, μ::DiscreteNonParametric, ν::DiscreteNonParametric, ε; regularization=false, kwargs...)
+
+Compute the Sinkhorn Divergence between finite discrete
+measures `μ` and `ν` with respect to a cost function `c`
+and entropic regularization parameter `ε`.
+
+The Sinkhorn Divergence is computed as:
+```math
+S_{c,ε}(μ,ν) := OT_{c,ε}(μ,ν) - \\frac{1}{2}(OT_{c,ε}(μ,μ) + OT_{c,ε}(ν,ν)),
+```
+where `OT_{c,ε}(μ,ν)`, `OT_{c,ε}(μ,μ)` and `OT_{c,ε}(ν,ν)` are the entropically
+regularized optimal transport cost between `(μ,ν)`, `(μ,μ)` and `(ν,ν)`, respectively.
+
+The formulation for the Sinkhorn Divergence may have slight variations depending on the paper consulted.
+The Sinkhorn Divergence was initially proposed by [^GPC18], although, this package uses the formulation given by
+[^FeydyP19], which is also the one used on the Python Optimal Transport package.
+
+[^GPC18]: Aude Genevay, Gabriel Peyré, Marco Cuturi, Learning Generative Models with Sinkhorn Divergences,
+Proceedings of the Twenty-First International Conference on Artficial Intelligence and Statistics, (AISTATS) 21, 2018
+
+[^FeydyP19]: Jean Feydy, Thibault Séjourné, François-Xavier Vialard, Shun-ichi
+Amari, Alain Trouvé, and Gabriel Peyré. Interpolating between op-
+timal transport and mmd using sinkhorn divergences. In The 22nd In-
+ternational Conference on Artificial Intelligence and Statistics, pages
+2681–2690. PMLR, 2019.
+
+See also: [`sinkhorn2`](@ref)
+"""
+function sinkhorn_divergence(
+    c,
```

Comment:
Could we implement this with the cost matrices as inputs instead? I think this would have multiple advantages:
- it would allow to reuse precomputed cost matrices, here and in other functions
- it would be guaranteed to work on the GPU (BTW GPU tests should be added if possible) whereas the default implementation of `pairwise` probably doesn't for most distances and custom functions (IIRC it uses incorrect containers and indexing which should be avoided on GPUs and is often disabled by users)

Reply:
Additionally, this would still allow to define the function implemented by just forwarding it to the one with cost matrices that are computed from the `DiscreteNonParametric` and the cost function.

Reply:
BTW it's also quite restrictive to only allow univariate messures here it seems.

Reply:
Yeah, the univariate was a mistake. I forgot that `DiscreteNonParametric` was only for the univariate case. I could implement it with a matrix cost, but it would actually require 3 matrices, so the it would be quite "ugly". It would be something like:
`sinkhorn_divergence(mu,nu,C1,C2,C3,eps)`. So I think we should stick with the cost function... I don't know much about programming for GPU, but, don't you think we could adapt this somehow?

Reply:
I could implement another `sinkhorn_divergence` that would take the cost matrices as argument. And would could perhaps indicate that such version should be used in case the user is interested in using GPU.

Reply:
You can just define both and forward the call from the one with the cost function to the other:

> Additionally, this would still allow to define the function implemented by just forwarding it to the one with cost matrices that are computed from the DiscreteNonParametric and the cost function.

Reply:
>  such version should be used in case the user is interested in using GPU.

It could be useful without GPUs as well, e.g., if you evaluate `sinkhorn`, `sinkhorn2` or other other regularized OT distances with the same cost matrices.

In fact, it could also be useful to allow to specify pre-computed plans for the different `sinkhorn2` calls.

Reply:
Convinced.